pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,383 @@
+		return *result, err
+	}
+
+	status := job.Status
+	if status.Succeeded > 0 {
+		return reconcile.Result{}, r.UpdatePhase(v2.BatchSucceeded, nil)
+	}
+
+	if status.Failed > 0 {
+		numConditions := len(status.Conditions)
+		if numConditions > 0 {
+			condition := status.Conditions[numConditions-1]
+
+			if condition.Type == batchv1.JobFailed {
+				var reason string
+				podName, err := r.getJobPodName(batch.Name, batch.Namespace)
+				if err != nil {
+					reason = err.Error()
+				} else {
+					reason, err = kubernetes.Logs(podName, batch.Namespace)
+					if err != nil {
+						reason = fmt.Errorf("Unable to retrive logs for batch %s: %w", batch.Name, err).Error()
+					}
+				}
+
+				_, err = r.update(func() {
+					r.instance.Status.Phase = v2.BatchFailed
+					r.instance.Status.Reason = reason
+				})
+				return reconcile.Result{}, err
+			}
+		}
+	}
+	// The job has not completed, wait 1 second before retrying
+	return reconcile.Result{}, nil
+}
+
+func (r *batchResource) cleanup() (reconcile.Result, error) {
+	batch := r.instance
+	meta := metav1.ObjectMeta{
+		Namespace: batch.Namespace,
+		Name:      batch.Name,
+	}
+	var allErrors error
+	// Delete the Job used to execute the batch
+	if err := r.client.Delete(ctx, &batchv1.Job{ObjectMeta: meta}); err != nil && !errors.IsNotFound(err) {
+		allErrors = fmt.Errorf("Unable to delete Job: %w", err)
+	}
+
+	// Delete the failed job pods used to execute the batch.
+	// Manual deletion is required as we use RestartPolicyNever on the job so that we can retrieve the cli logs
+	// in the event of a failure
+	if batch.Status.Phase == v2.BatchFailed {
+		labels := batchLabels(batch.Name)
+		r.client.DeleteAllOf(ctx, &corev1.Pod{}, []client.DeleteAllOfOption{
+			client.InNamespace(batch.Namespace),
+			client.MatchingLabels(labels),
+		}...)
+	}
+	return reconcile.Result{}, allErrors
+}
+
+func connectionUrl(i *v1.Infinispan) (string, error) {
+	var authString string
+	if i.IsAuthenticationEnabled() {
+		user := consts.DefaultOperatorUser
+		pass, err := users.PasswordFromSecret(user, i.GetSecretName(), i.GetNamespace(), kubernetes)
+		if err != nil {
+			return "", err
+		}
+		authString = fmt.Sprintf("%s:%s@", user, pass)
+	}
+
+	url := fmt.Sprintf("%s://%s%s:%d", i.GetEndpointScheme(), authString, i.GetServiceFqn(), consts.InfinispanPort)
+	return url, nil
+}
+
+func (r *batchResource) loadCluster() (*v1.Infinispan, error) {
+	batch := r.instance
+	infinispan := &v1.Infinispan{}
+	clusterName := batch.Spec.Cluster
+	clusterObjKey := types.NamespacedName{
+		Namespace: batch.Namespace,
+		Name:      clusterName,
+	}
+	if err := r.client.Get(ctx, clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, fmt.Errorf("Infinispan CR '%s' not found", clusterName)
+		}
+		// Error reading the object - requeue the request.
+		return nil, fmt.Errorf("Unable to fetch Infinispan CR '%s': %w", clusterName, err)
+	}
+	return infinispan, nil
```

Comment:
Seems to be `LookupResource` could be utilized here

Reply:
Much better. This allowed me to remove the `loadCluster` function entirely.